tiledb/sm/query/readers/sparse_index_reader_base.h

```diff
@@ -290,6 +304,186 @@ class SparseIndexReaderBase : public ReaderBase {
     bool done_adding_result_tiles_;
   };
 
+  /**
+   * Temporary read state that can be recomputed using just the read state. This
+   * contains information about the tile ranges that still need to be processed
+   * if a subarray is set (with memory used) and for which fragments we loaded
+   * all tiles into memory. It also contains which tiles contains no results so
+   * they should be ignored by further iterations. It should be used in
+   * conjunction with the loaded tiles in the respective readers. Eventually,
+   * those will also move to this class but that requires the removal of
+   * every memory allocated things from result tiles into separate objects so
+   * that the list of result tiles is only a fragment/tile index and the objects
+   * are not different anymore.
+   */
+  class TransientReadState : public ITileRange {
+   public:
+    /* ********************************* */
+    /*     CONSTRUCTORS & DESTRUCTORS    */
+    /* ********************************* */
+
+    TransientReadState() = delete;
+
+    TransientReadState(uint64_t num_frags)
+        : tile_ranges_(num_frags)
+        , memory_used_tile_ranges_(0)
+        , all_tiles_loaded_(num_frags, false) {
+    }
+
+    DISABLE_COPY_AND_COPY_ASSIGN(TransientReadState);
+    DISABLE_MOVE_AND_MOVE_ASSIGN(TransientReadState);
+
+    /* ********************************* */
+    /*          PUBLIC METHODS           */
+    /* ********************************* */
+
+    /**
+     * Return the tile ranges vector for a particular fragment.
+     *
+     * @param f Fragment index.
+     * @return Tile ranges.
+     */
+    std::vector<std::pair<uint64_t, uint64_t>>& tile_ranges(const unsigned f) {
```

Comment:
We are trying to move toward named types. In this case it might be clearer (and shorter?) to change this from struct-of-arrays to array-of-structs, where the range pair, memory used, and loaded flag are bundled into a POD struct.

Reply:
I think the right design here is to move this to an object, and also move all the computations for tile ranges that are in the subarray to that object. This was a little too much work for this refactor so I filed https://app.shortcut.com/tiledb-inc/story/28389/make-object-for-tile-ranges-in-sparse-index-readers.